File: _Basic/windows-forms-introduction-with-visual-basic/8-windows-forms-introduction-with-visual-basic-m8-linq-exercise-files/CS/Data/Data/DataForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace Data
{
    public partial class DataForm : Form
    {
        public DataForm()
        {
            InitializeComponent();
        }

        private BindingSource _categoriesBindingSource = new BindingSource();
        private BindingSource _productsBindingSource = new BindingSource();

        private ISource _currentSource;
        private ObjectSource _objectSource;
        private DataSetSource _dataSetSource;
        private LinqSource _linqSource;

        private void DataForm_Load(object sender, EventArgs e)
        {
            try
            {
                SourceToolStripComboBox.SelectedIndex = 0;
                SetSource();
                BindCategories();

                _categoriesBindingSource.DataSource = _currentSource.GetCategories();

                CategoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName";
                CategoryToolStripComboBox.ComboBox.ValueMember = "CategoryID";
                CategoryToolStripComboBox.ComboBox.DataSource = _categoriesBindingSource;

                ProductsDataGridView.DataSource = _productsBindingSource;
                ProductsListBox.DisplayMember = "ProductName";
                ProductsListBox.ValueMember = "ProductID";
                ProductsListBox.DataSource = _productsBindingSource;

                NameTextBox.DataBindings.Add("Text", _productsBindingSource, "ProductName", true);
                QuantityTextBox.DataBindings.Add("Text", _productsBindingSource, "QuantityPerUnit", true);
                PriceTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitPrice", true);
                StockTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsInStock", true);
                OrderTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsOnOrder", true);
                DiscontinuedCheckBox.DataBindings.Add("Checked", _productsBindingSource, "Discontinued", true);
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }

        private void SourceToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            SetSource();
            BindCategories();
            BindProducts();
        }

        private void CategoryToolStripComboBox_SelectedIndexChanged(object sender, EventArgs e)
        {
            BindProducts();
        }

        private void SetSource()
        {
            switch (SourceToolStripComboBox.SelectedIndex)
            {
                case 0:
                    if (_objectSource == null)
                    {
                        _objectSource = new ObjectSource();
                    }
                    _currentSource = _objectSource;
                    break;
                case 1:
                    if (_dataSetSource == null)
                    {
                        _dataSetSource = new DataSetSource();
                    }
                    _currentSource = _dataSetSource;
                    break;
                case 2:
                    if (_linqSource == null)
                    {
                        _linqSource = new LinqSource();
                    }
                    _currentSource = _linqSource;
                    break;
            }
        }

        private void BindCategories()
        {
            _categoriesBindingSource.DataSource = _currentSource.GetCategories();
        }

        private void BindProducts()
        {
            var catId = Convert.ToInt32(CategoryToolStripComboBox.ComboBox.SelectedValue);
            _productsBindingSource.DataSource = _currentSource.GetProducts(catId);
        }

        private void AddToolStripButton_Click(object sender, EventArgs e)
        {
            var id = Convert.ToInt32(CategoryToolStripComboBox.ComboBox.SelectedValue);
            var name = CategoryToolStripComboBox.Text;
            var category = new Category(id, name);

            AddProductForm form = new AddProductForm(category);
            var result = form.ShowDialog();

            if (result == DialogResult.OK)
            {
                var rebind = _currentSource.AddProduct(form.Product);
                if (rebind)
                {
                    BindProducts();
                }
            }
        }

        private void DeleteToolStripButton_Click(object sender, EventArgs e)
        {
            var productId = Convert.ToInt32(ProductsListBox.SelectedValue);
            var rebind = _currentSource.DeleteProduct(productId);
            if (rebind)
            {
                BindProducts();
            }
        }

        private void SaveToolStripButton_Click(object sender, EventArgs e)
        {
            _currentSource.Save();
        }

        private void BackToolStripButton_Click(object sender, EventArgs e)
        {
            _productsBindingSource.MovePrevious();
        }

        private void ForwardToolStripButton_Click(object sender, EventArgs e)
        {
            _productsBindingSource.MoveNext();
        }
    }
}

File: _Basic/windows-forms-introduction-with-visual-basic/6-windows-forms-introduction-with-visual-basic-m6-data-binding-exercise-files/VB/Data/Data/DataForm.vb
Public Class DataForm
    Private _source As New ObjectSource()
    Private _categoriesBindingSource As New BindingSource()
    Private _productsBindingSource As New BindingSource()

    Private Sub DataForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        _categoriesBindingSource.DataSource = _source.GetCategories()

        CategoryToolStripComboBox.ComboBox.DisplayMember = "CategoryName"
        CategoryToolStripComboBox.ComboBox.ValueMember = "CategoryID"
        CategoryToolStripComboBox.ComboBox.DataSource = _categoriesBindingSource

        ProductsDataGridView.DataSource = _productsBindingSource
        ProductsListBox.DataSource = _productsBindingSource
        ProductsListBox.DisplayMember = "ProductName"

        NameTextBox.DataBindings.Add("Text", _productsBindingSource, "ProductName")
        QuantityTextBox.DataBindings.Add("Text", _productsBindingSource, "QuantityPerUnit")
        PriceTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitPrice")
        StockTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsInStock")
        OrderTextBox.DataBindings.Add("Text", _productsBindingSource, "UnitsOnOrder")
        DiscontinuedCheckBox.DataBindings.Add("Checked", _productsBindingSource, "Discontinued")
    End Sub

    Private Sub CategoryToolStripComboBox_SelectedIndexChanged(sender As Object, e As EventArgs) Handles CategoryToolStripComboBox.SelectedIndexChanged
        Dim catId = CategoryToolStripComboBox.ComboBox.SelectedValue
        _productsBindingSource.DataSource = _source.GetProducts(catId)
    End Sub

    Private Sub AddToolStripButton_Click(sender As Object, e As EventArgs) Handles AddToolStripButton.Click
        Dim category = CType(CategoryToolStripComboBox.SelectedItem, Category)
        Dim form As New AddProductForm(category)
        Dim result = form.ShowDialog()

        If result = Windows.Forms.DialogResult.OK Then
            ' _source.AddProduct(form.Product)
            _productsBindingSource.Add(form.Product)
        End If
    End Sub

    Private Sub DeleteToolStripButton_Click(sender As Object, e As EventArgs) Handles DeleteToolStripButton.Click
        Dim product = CType(ProductsListBox.SelectedItem, Product)
        ' _source.DeleteProduct(product)
        _productsBindingSource.Remove(product)
    End Sub

    Private Sub BackToolStripButton_Click(sender As Object, e As EventArgs) Handles BackToolStripButton.Click
        _productsBindingSource.MovePrevious()
    End Sub

    Private Sub ForwardToolStripButton_Click(sender As Object, e As EventArgs) Handles ForwardToolStripButton.Click
        _productsBindingSource.MoveNext()
    End Sub

End Class
